Add RequiresSpecificOwner trait and lint check

reaperrr opened the pull request:
The purpose of this trait is to allow Lint and the map editor to check for and enforce specific owner(s) for actors where any other owner could cause issues (case in point: #15457).

In practice, this is only used to enforce Neutral on RA/TD trees for now, because their Targetable trait makes AI squads consider them as valid targets when owned by the 'enemy' Creeps.
The fact that the AI only looks for the Targetable trait and not whether the target type is actually valid for the squad leader is certainly another issue, but this PR has its own merits regardless of that.

reaperrr committed:
Add RequiresSpecificOwners trait

To enforce specific owners via Lint rules,
and possibly other means later.
This is for cases where accidentally setting an
unfitting owner via editor could cause issues.

Example: AI might try to attack Creeps-owned trees
and get stuck.

reaperrr committed:
Enforce required owner in map editor

It can easily happen that mappers forget to set the
current player to Neutral before placing more trees,
for example, so we force the editor to set a valid owner.

pchote commented on OpenRA.Mods.Common/Traits/RequiresSpecialOwner.cs:
```diff
@@ -0,0 +1,43 @@
+	public class RequiresSpecialOwnerInfo : ITraitInfo
+	{
+		[Desc("Owner to be enforced via Lint rule and map editor. Currently, the only options are Neutral and Creeps.")]
+		public readonly SpecialOwnerType ValidOwnerTypes = SpecialOwnerType.Neutral;
```
Why not support strings here?

pchote replied:
IMO this should also be a required value instead of defaulting to neutral.

reaperrr commented:
Updated and rebased.

abcdefg30 commented on OpenRA.Mods.Common/EditorBrushes/EditorActorBrush.cs:
```diff
@@ -50,10 +51,15 @@ public EditorActorBrush(EditorViewportControllerWidget editorWidget, ActorInfo a
+			// Enforce Neutral or Creeps owner if the actor has RequiresSpecificOwner
+			var specificOwnerInfo = actor.TraitInfoOrDefault<RequiresSpecificOwnerInfo>();
+			if (specificOwnerInfo != null && !specificOwnerInfo.ValidOwnerTypes.Contains(ownerName))
+				ownerName = specificOwnerInfo.ValidOwnerTypes.First();
```
I wonder if it is worth to log this (in e.g. `debug.log`). Or even throw an exception instead of silently failing.

Mailaender replied:
This is silently correcting the error. An exception could mean loosing all mapping progress, which would disturb things way too much.

Mailaender reviewed (dismissed):
Looking good.

pchote commented on OpenRA.Mods.Common/Lint/CheckPlayers.cs:
```diff
@@ -89,6 +89,20 @@ public void Run(Action<string> emitError, Action<string> emitWarning, ModData mo
 					}
+
+					// Some actors may have RequiresSpecificOwner to enforce a specific owner (example: 'Neutral' for trees)
+					foreach (var actor in map.Rules.Actors)
```
Why are we enumerating all actors? We already know exactly what actor we want to look up here!

pchote commented:
What's the status here @reaperrr?

reaperrr commented:
> 1. Query the rules for all actors with the `RequiresSpecificOwnerInfo` traitinfo
2. Build a dictionary of actor name &rarr; required owner
3. Enumerate actor definitions
4. If actor type exists in the dictionary parse the owner and check that it matches

None of these steps are are so 100% clear to me that I wouldn't first have to get my head around them and look for probable examples, and then trial and error my way through them. That's always a somewhat time-consuming and tiresome process, so I haven't found the motivation for it yet.

pchote commented:
Ok. I can put a prototype together for that if you'd like.

pchote commented:
fc0cc214819580cd28bafd7a9a1b44f033117b2a

reaperrr committed:
Remove hardcoded mpspawn owner lint check

Use RequiresSpecificOwner to enforce the owner
that owns the world instead.

Require 'Neutral' in the official mods accordingly.

reaperrr commented:
Updated with ^ prototype (with tiny adjustment to lint error message, and removal of hardcoded `mpspawn` check split to separate commit as discussed on IRC).

pchote commented on mods/ra/rules/defaults.yaml:
```diff
@@ -892,6 +892,8 @@
 		ExcludeTilesets: INTERIOR
 		Categories: Tree
+	RequiresSpecificOwners:
+		ValidOwnerNames: Neutral
```
We'll also want this on the burnt tree husks, and maybe other actors.  We have mpspawn as a testcase so it might be worthwhile splitting the "normal" rules for this into a followup PR where we can bikeshed all the extra actor types that should have this?

pchote commented on OpenRA.Mods.Common/Traits/RequiresSpecificOwners.cs:
```diff
@@ -0,0 +1,37 @@
+
+	public class RequiresSpecificOwners
+	{
+		public readonly RequiresSpecificOwnersInfo Info;
```
This isn't used, so the whole thing can simplify down to

```csharp
using System.Collections.Generic;
using OpenRA.Traits;

namespace OpenRA.Mods.Common.Traits
{
	[Desc("Can be used to enforce specific owners (like 'Neutral' or 'Creeps') for this actor.")]
	public class RequiresSpecificOwnersInfo : TraitInfo<RequiresSpecificOwners>
	{
		[Desc("Only allow players listed here as owners.")]
		[FieldLoader.Require]
		public readonly HashSet<string> ValidOwnerNames = new HashSet<string>();
	}

	public class RequiresSpecificOwners { }
}
```

pchote commented:
https://github.com/OpenRA/OpenRA/wiki/Changelog-(bleed)/_compare/c991f925697032700f7f01c7d6bb094a5dc5bca9

pchote commented:
https://github.com/OpenRA/OpenRA/wiki/Changelog-(bleed)/_compare/0a79aea7b5c6aece016fba557f9777b80329e335
